src/main/java/uk/gov/ons/ctp/response/notify/util/BitCalculator.java

```diff
@@ -0,0 +1,207 @@
+         * @param aValue Value represented by input string
+         * @param aValueStr Value portion of input string
+         * @param aKeyName Name of a key
+         * @param aBits Number of bits represented by data in input string
+         * @param aRadix Radix of number in input string
+         * @param aValid Validity flag
+         */
+        public KeyInfo(
+                final String aInputString,
+                final BigInteger aValue,
+                final String aValueStr,
+                final String aKeyName,
+                final int aBits,
+                final int aRadix,
+                final boolean aValid) {
+            this.inputString = aInputString;
+            this.value = aValue;
+            this.valueStr = aValueStr;
+            this.keyName = aKeyName;
+            this.bits = aBits;
+            this.radix = aRadix;
+            this.valid = aValid;
+        }
+
+        @Override
+        public String toString() {
+            // Extremely verbose format mandated by checkstyle
+            return "KeyInfo{"
+                    +
+                    "inputString='"
+                    +
+                    inputString
+                    +
+                    '\''
+                    +
+                    ", value="
+                    +
+                    value
+                    +
+                    ", valueStr='"
+                    +
+                    valueStr
+                    +
+                    '\''
+                    +
+                    ", keyName='"
+                    +
+                    keyName
+                    +
+                    '\''
+                    +
+                    ", bits="
+                    +
+                    bits
+                    +
+                    ", radix="
+                    +
+                    radix
+                    +
+                    '}';
+        }
+    }
+
+    /**
+     * Parses a String into a KeyInfo structure.  If the isKey flag is set the String is a gov.notify key and the text
+     * up to the first hyphen is assumed to be the name of the key.  If the isKey flag is not set, the string is a
+     * number in base radix with optional hyphens to break it up.
+     * @param str string to parse
+     * @param isKey key flag to indicate whether the string is a gov.notify key
+     * @param radix the radix of the number portion
+     * @return a KeyInfo object showing the number of bits of data contained within the number portion of the supplied string
+     */
+    public KeyInfo analyse(final String str, final boolean isKey, final int radix) {
+        KeyInfo info = cleanString(str, isKey, radix);
+
+        if (info.valid) {
+            int numDigits = info.valueStr.length();
+            int bitsPerDigit = new Double(getBitsPerDigit(radix)).intValue();
+            int bits = ((numDigits * bitsPerDigit) >>> radix);
+            boolean valid = isKey == false || bits >= MINIMUM_KEY_BITS;
```

Comment:
boolean valid = !isKey || bits >= MINIMUM_KEY_BITS

Reply:
I disagree - use of ! is easily misread.   We can agree to differ :-)

Reply:
I've put parentheses around it

Reply:
Cool, we'll agree to disagree on this one then